File: hmm_tmb.R
# HMM with TMB
library(TMB)
compile("hmm_tmb.cpp")
dyn.load(dynlib("hmm_tmb"))

SimulatePoHmm <- function(n, lambda, tpm, n.states, delta) {
 s <- numeric(n)
 state.space <- 1:n.states
 s <- sample(state.space, 1, prob = delta)
 for (t in 2:n) s[t] <- sample(state.space, 1, prob = tpm[s[t - 1], ])
 data <- rpois(n, lambda = lambda[s])
 return(data)
}

ConvertN2W <- function(lambda, tpm, n.states) { 
  # Poisson means first, log link 
  wpar <- log(c(lambda[1], diff(lambda)))
  # Transition probabilities, logit link
  tr_tpm <- log(tpm / diag(tpm))
  wpar <- c(wpar, as.vector(tr_tpm[!diag(n.states)]))
  return(wpar)
}

ConvertW2N <- function(wpar, n.states) {
  # Poisson means first, log link
  lambda <- exp(wpar[1:n.states])
  lambda <- cumsum(lambda)
  # Tpm, logit link
  tpm <- diag(n.states)
  tpm[!diag(n.states)] <- exp(wpar[-(1:n.states)])
  tpm <- tpm / apply(tpm, 1, sum)
  delta <- solve(t(diag(n.states) - tpm  + 1), rep(1, n.states))
  return(list(lambda = lambda, tpm = tpm, delta = delta))
}

FitPoHmm <- function(data, n.states, ini.lambda, ini.tpm) {
  dat <- list(data = data, n_states = n.states)
  par <- list(wpar = ConvertN2W(ini.lambda, ini.tpm, n.states))
  obj <- MakeADFun(data = dat, parameters = par, DLL = "hmm_tmb", silent = TRUE)
  obj$hessian <- FALSE
  mod <- do.call("optim", obj)
  est <- ConvertW2N(mod$par, n.states)
  sd <- sdreport(obj)  
  llk <- -mod$value
  return(list(llk = llk, est = est, sd = sd))
}

File: hmm_tmb.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  // Read in data 
  DATA_VECTOR(data);
  DATA_INTEGER(n_states);
  // Read in parameters
  PARAMETER_VECTOR(wpar);
  // Unpack and transform parameters
  vector<Type> lambda(wpar.head(n_states));
  lambda = exp(lambda);
  for (int i = 0; i < n_states - 1; ++i) lambda(i + 1) += lambda(i);  
  matrix<Type> tpm(n_states, n_states);
  int cur = n_states; 
  for (int i = 0; i < n_states; ++i) {
    tpm(i, i) = 1; 
   for (int j = 0; j < n_states; ++j) {
    if (i != j) {
      tpm(i, j) = exp(wpar[cur]); 
      ++cur; 
    } 
   } 
   tpm.row(i) /= tpm.row(i).sum();
  } 
  // Compute stationary distribution 
  matrix<Type> I = matrix<Type>::Identity(n_states, n_states);
  matrix<Type> tpminv = I; 
  tpminv -= tpm; 
  tpminv = (tpminv.array() + 1).matrix(); 
  matrix<Type> ivec(1, n_states); for (int i = 0; i < n_states; ++i) ivec(0, i) = 1; 
  tpminv = tpminv.inverse();
  matrix<Type> delta = ivec * tpminv;
  // compute observation probabilities
  int n = data.rows();  
  matrix<Type> prob(n, n_states); 
  for (int s = 0; s < n_states; ++s) {
    for (int i = 0; i < n; ++i) {
      prob(i, s) = dpois(data(i), lambda[s]); 
    }
  } 
  // compute log-likelihood 
  Type llk = 0;
  matrix<Type> phi(delta);
  Type sumphi = 0;
  for (int i = 0; i < n; ++i) {
    phi = (phi.array() * prob.row(i).array()).matrix();  
    phi = phi * tpm;
    sumphi = phi.sum();
    llk += log(sumphi);
    phi /= sumphi;
  }
  Type nll = -llk; 
  //std::cout << "llk: " << llk << std::endl; 
  return nll;
}
